### PR TITLE
Support scalar subqueries cache 

### PR DESCRIPTION
**Pull request by SmitaRKulkarni**
Changelog category (leave one):
* Improvements

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support scalar subqueries cache 
Implementation:
* Added a map with hash of the node (without alias) and the evaluated value to Context.
Testing:
* Added a test-case with new analyser in 02174_cte_scalar_cache.sql

### CODE REVIEW

**Comment by Algunenano** [comment]
I guess that this being based on query hash, which currently ignores identifier aliases, means the same bugs found in the current interpreter will affect the analyzer. See https://github.com/ClickHouse/ClickHouse/pull/31796#issuecomment-1273009107 or other tests in https://github.com/ClickHouse/ClickHouse/pull/40065/

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 1]
```diff
@@ -66,6 +66,15 @@
 #include <Analyzer/UnionNode.h>
 #include <Analyzer/InDepthQueryTreeVisitor.h>
 #include <Analyzer/QueryTreeBuilder.h>
+#include <Analyzer/IQueryTreeNode.h>
+#include <Common/ProfileEvents.h>
+#include <Analyzer/HashUtils.h>
+
+namespace ProfileEvents
+{
+extern const Event ScalarSubqueriesGlobalCacheHit;
```
Comment (kitaisreal) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:75]: Need to add intendation with 4 spaces.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 2]
```diff
@@ -66,6 +66,15 @@
 #include <Analyzer/UnionNode.h>
 #include <Analyzer/InDepthQueryTreeVisitor.h>
 #include <Analyzer/QueryTreeBuilder.h>
+#include <Analyzer/IQueryTreeNode.h>
+#include <Common/ProfileEvents.h>
```
Comment (kitaisreal) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:70]: We include files from `Common` at the top of the file.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 3]
```diff
@@ -1682,11 +1694,22 @@ void QueryAnalyzer::evaluateScalarSubqueryIfNeeded(QueryTreeNodePtr & node, size
     auto * query_node = node->as<QueryNode>();
     auto * union_node = node->as<UnionNode>();
     if (!query_node && !union_node)
-        throw Exception(ErrorCodes::LOGICAL_ERROR,
+        throw Exception(
```
Comment (kitaisreal) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:1697]: This does not look valid or better readable than before. Consider to revert this line.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 4]
```diff
@@ -1717,7 +1743,8 @@ void QueryAnalyzer::evaluateScalarSubqueryIfNeeded(QueryTreeNodePtr & node, size
         if (!type->isNullable())
         {
             if (!type->canBeInsideNullable())
-                throw Exception(ErrorCodes::INCORRECT_RESULT_OF_SCALAR_SUBQUERY,
+                throw Exception(
```
Comment (kitaisreal) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:1746]: This does not look valid or better readable than before. Consider to revert this line.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 5]
```diff
@@ -1699,10 +1722,13 @@ void QueryAnalyzer::evaluateScalarSubqueryIfNeeded(QueryTreeNodePtr & node, size
 
     auto io = interpreter->execute();
 
-    Block block;
     PullingAsyncPipelineExecutor executor(io.pipeline);
     io.pipeline.setProgressCallback(context->getProgressCallback());
 
+    Block block;
+    Field scalar_value;
```
Comment (kitaisreal) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:1729]: Not sure why we need to move `scalar_value` and `scalar_type` at the top of the function.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 6]
```diff
@@ -1738,15 +1765,11 @@ void QueryAnalyzer::evaluateScalarSubqueryIfNeeded(QueryTreeNodePtr & node, size
     }
 
     if (tmp_block.rows() != 0)
-        throw Exception(ErrorCodes::INCORRECT_RESULT_OF_SCALAR_SUBQUERY, "Scalar subquery returned more than one row");
+        throw Exception(ErrorCodes::INCORRECT_RESULT_OF_SCALAR_SUBQUERY, "Scalar subquery returned more than zero row");
```
Comment (kitaisreal) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:1768]: The exception was right. We check this `tmp_block` after subquery already returned one row.

File: tests/queries/0_stateless/02174_cte_scalar_cache.sql [hunk 7]
```diff
@@ -18,6 +18,16 @@ SELECT '02177_CTE_GLOBAL_OFF', a5 FROM system.numbers LIMIT 100
         FORMAT Null
 SETTINGS enable_global_with_statement = 0;
 
+WITH
```
Comment (kitaisreal) [inline tests/queries/0_stateless/02174_cte_scalar_cache.sql:21]: This test looks strange, why the result is different with `allow_experimental_analyzer` from `02177_CTE_GLOBAL_ON` or `02177_CTE_GLOBAL_OFF`.

File: tests/queries/0_stateless/02174_cte_scalar_cache.sql [hunk 8]
```diff
@@ -46,3 +56,17 @@ WHERE
   AND type = 'QueryFinish'
   AND query LIKE '%02177_CTE_GLOBAL_OFF%'
   AND event_date >= yesterday() AND event_time > now() - interval 10 minute;
+
+SELECT
```
Comment (kitaisreal) [inline tests/queries/0_stateless/02174_cte_scalar_cache.sql:60]: Need to add test specifically for analyzer scalar subqueries.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Analyzer/IQueryTreeNode.cpp [hunk 9]
```diff
@@ -9,6 +9,7 @@
 #include <IO/Operators.h>
 
 #include <Parsers/ASTWithAlias.h>
+#include <Analyzer/QueryNode.h>
```
[the remark this answers is not in the document]
Comment (SmitaRKulkarni) [reply src/Analyzer/IQueryTreeNode.cpp]: This was actually used for previous approach while editing equal function. No longer needed. Removed it.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp (hunk 1 above)
Comment (SmitaRKulkarni) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp]: addressed it

File: src/Analyzer/Passes/QueryAnalysisPass.cpp (hunk 2 above)
Comment (SmitaRKulkarni) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp]: addressed it

File: src/Analyzer/Passes/QueryAnalysisPass.cpp (hunk 3 above)
Comment (SmitaRKulkarni) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp]: Unintentional change, reverted it

File: src/Analyzer/Passes/QueryAnalysisPass.cpp (hunk 4 above)
Comment (SmitaRKulkarni) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp]: Unintentional change, reverted it

File: src/Analyzer/Passes/QueryAnalysisPass.cpp (hunk 5 above)
Comment (SmitaRKulkarni) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp]: I had moved it above for the previous approach, but now no longer needed. Reverted it.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp (hunk 6 above)
Comment (SmitaRKulkarni) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp]: Thanks for the clarification. Reverted it. Its quite confusing as the check is for `rows != 0`

File: tests/queries/0_stateless/02174_cte_scalar_cache.sql (hunk 7 above)
Comment (SmitaRKulkarni) [reply tests/queries/0_stateless/02174_cte_scalar_cache.sql:21]: It should be same as 02177_CTE_GLOBAL_ON, the only difference with alias are not considered as same queries, thus we have one more cache miss.

File: tests/queries/0_stateless/02174_cte_scalar_cache.sql (hunk 8 above)
Comment (SmitaRKulkarni) [reply tests/queries/0_stateless/02174_cte_scalar_cache.sql:60]: This test with '02177_CTE_NEW_ANALYZER', is for analyser scalar subqueries with `SELECT sleep(0.0001) FROM system.one` being the scalar subquery. 
If you have any other queries which should be added to the test, could you please share some examples.

**Review by kitaisreal — APPROVED** [review]

**Comment by kitaisreal** [comment]
@SmitaRKulkarni let's wait for tests to pass and merge.

**Pull request merged**